text/0000-pin_and_move.md

```diff
@@ -0,0 +1,381 @@
+  This does not leave forward compatibility for the static string optimization
+  mentioned previously.
+- That crate has no equivalent to the `Own` trait, which is necessary for some
+  APIs using internal references.
+- That crate has a `StableDerefClone` type, which is equivalent to the bound
+  `Share + StableDeref` in our system.
+
+If the hierarchy proposed in this RFC becomes stable, all users are encouraged
+to migrate from that crate to the standard library traits.
+
+## The `Pin` type
+
+The `Pin` type is a wrapper around a mutable reference. If the type it
+references is `!Move`, the `Pin` type guarantees that the referenced data will
+never be moved again. It has a relatively small API. It is added to both
+`std::mem` and `core::mem`.
+
+```rust
+struct Pin<'a, T: ?Sized + 'a> {
+    data: &'a mut T,
+}
+
+impl<'a, T: ?Sized + Move> Pin<'a, T> {
+    pub fn new(data: &'a mut T) -> Pin<'a, T>;
+}
+
+impl<'a, T: ?Sized> Pin<'a, T> {
+    pub unsafe fn new_unchecked(data: &'a mut T) -> Pin<'a, T>;
+
+    pub unsafe fn get_mut(this: Pin<'a, T>) -> &'a mut T;
+
+    pub fn borrow<'b>(this: &'b mut Pin<'a, T>) -> Pin<'b, T>;
+}
+
+impl<'a, T: ?Sized> Deref for Pin<'a, T> {
+    type Target = T;
+}
+
+impl<'a, T: ?Sized + Move> DerefMut for Pin<'a, T> { }
+```
+
+For types which implement `Move`, `Pin` is essentially the same as an `&mut T`.
+But for types which do not, the conversion between `&mut T` and `Pin` is
+unsafe (however, `Pin` can be easily immutably dereferenced, even for `!Move`
+types).
+
+The contract on the unsafe part of `Pin`s API is that a Pin cannot be
+constructed if the data it references would ever move again, and that it cannot
+be converted into a mutable reference if the data might ever be moved out of
+that reference. In other words, if you have a `Pin` containing data which does
+not implement `Move`, you have a guarantee that that data will never move.
+
+The next two subsections describe safe APIs for constructing a `Pin` of data
+which cannot be moved - one in the heap, and one in the stack.
+
+### Pinning to the heap: The `Anchor` type
+
+The `Anchor` wrapper takes a type that implements `StableDeref` and `Own`, and
+prevents users from moving data out of that unless it implements `Move`. It is
+added to `std::mem` and `core::mem`.
+
+```rust
+struct Anchor<T> {
+     ptr: T,
+}
+
+impl<T: StableDerefMut + Own> Anchor<T> {
+     pub fn new(ptr: T) -> Anchor<T>;
+
+     pub unsafe fn get_mut(this: &mut Anchor<T>) -> &mut T;
+
+     pub unsafe fn into_inner_unchecked(this: Anchor<T>) -> T;
+
+     pub fn pin<'a>(this: &'a mut Anchor<T>) -> Pin<'a, T::Target>;
+}
+
+impl<T: StableDerefMut + Own> Anchor<T> where T::Target: Move {
+     pub fn into_inner(this: Anchor<T>) -> T;
+}
+
+impl<T: StableDerefMut + Own> Deref for Anchor<T> {
+     type Target = T;
+}
+
+impl<T: StableDerefMut + Own> DerefMut for Anchor<T> where T::Target: Move { }
+```
```

Comment:
Should there be an `unsafe impl<T: StableDerefMut + Own> Move for Anchor {}`?

Reply:
I don't think so. It's conceivable that a smart pointer could itself be `!Move`.

Because the smart pointers themselves implement `Move` even if their referent doesn't, something like `Anchor<Box<T: !Move>>` should implement `Move`.

Reply:
Ah, so is `Anchor` going to be `#[fundamental]` to allow for `Move` impls for `Anchor<MyBox<T>>` for custom `MyBox`es?

Reply:
You should just implement Move for `MyBox`, `T: Move` implies `Anchor<T>: Move`. I don't know how `Anchor<T>` could be `Move` if `T` wasn't, since its just a newtype.

Reply:
> You should just implement Move for MyBox

I can't, even for `Box`. `Box<MyImmovableGenerator>` is still `!Move`.

> I don't know how Anchor<T> could be Move if T wasn't, since its just a newtype.

`Box<MyImmovableGenerator>` is `!Move`, but `Anchor<Box<MyImmovableGenerator>>` is `Move`, right?

In order to make this work, you need something like `unsafe impl<T> Move for Anchor<Box<T>> {}`. In order for that to be coherent for external crates which define `MyBox`, you'd need `Anchor` to be `#[fundamental]`

Reply:
Sorry, I'm now rereading your earlier comment:
> Because the smart pointers themselves implement Move even if their referent doesn't

That can't be true-- if it was, I could write the following:

```rust
let mut movable = Box::new(immovable_generator);
{
    Pin::new(&mut movable).poll(cx);
}
let movable_moved_to_stack = *movable;
let mut movable_moved_to_other_part_of_heap = Box::new(movable_moved_to_stack);
{
    Pin::new(&mut movable).poll(cx);
}
```

There can't be a `Move` impl until `Anchor` has been introduced into the picture.

Reply:
In your code, you call poll on a `Pin<'a, Box<impl Generator>>`. This relies on there being an impl of Future for `Box<impl Generator>`. I imagine this is going through the `F: Future => Box<F>: Future` impl. But that impl hasn't accounted for immovable types, and should be constrained to `F: Future + Move => Box<F>: Future`.

Reply:
I thought `Move` itself only meant "*moving* this type doesn't move anything that shouldn't be moved" - which is true for smart pointers regardless of their contents? And that it doesn't say anything about what operations on that type, such as deref, might do.

> I imagine this is going through the `F: Future => Box<F>: Future impl`. But that impl hasn't accounted for immovable types, and should be constrained to `F: Future + Move => Box<F>: Future`.

This confused me for a moment and I think I might've figured it out, but I'll write it down anyway. The plan is that an immovable generator, after combinators and so on had been applied to it, would be put into a single heap allocation (e.g. a `Box`) to be run. In this case clearly `F: !Move`, so `Box<F>: !Future`, which sounds bad. But, it's not a problem because it's going to be run by derefing the `Box` into an `&mut` and pinning it, rather than through the `Box<F>: Future` impl directly?

Reply:
@glaebhoerl 
> I thought `Move` itself only meant "moving this type doesn't move anything that shouldn't be moved"

My understanding is that `Move` means "functions that require `Pin<Self>` don't stop this type from being able to move in the future". `Box<T>` can't be `Move` without a `T: Move` bound because you could call `poll`, move the `T` out of the box (moving it), and then rebox it (in a different location) and re-poll it. This breaks the rule that `T: !Move` cannot be polled, moved, and then re-polled.

@withoutboats 
> In your code, you call poll on a Pin<'a, Box<impl Generator>>. This relies on there being an impl of Future for Box<impl Generator>. I imagine this is going through the F: Future => Box<F>: Future impl. But that impl hasn't accounted for immovable types, and should be constrained to F: Future + Move => Box<F>: Future.

You're saying that there's no longer an `impl<T: Future + ?Sized> Future for Box<T>`? I don't understand how that solves the problem. We want to have `Anchor<Box<Future>>: Future + Move`, right?

Reply:
> `Box<T>` can't be `Move` without a `T: Move` bound because

It says in the RFC that

>     unsafe impl<'a, T: ?Sized> Move for Box<T> { }

with the motivation:

> Positive impls of `Move` are added for types which contain pointers to generic types, but do not contain those types in their stack representation

Reply:
@withoutboats Can you clarify this point?